### PR TITLE
Release 1.0

### PR DESCRIPTION
**Pull request by BagdanGilevich**

### CODE REVIEW

**Comment by ilyahumbar** [comment]
Try to make granular git commits. It would be much easier to read changes and track history if you divided your task into subtasks. e.g.
- Add jasmine to the project
- Write unit tests for the first task
- Implement first task
- etc.
Subtasks might differ but the concept should stay the same.

File: week2-javascript/jasmine-standalone-2.4.1/spec/09First.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
+describe("Sum of random numbers", function () {
```
Comment (ilyahumbar) [inline week2-javascript/jasmine-standalone-2.4.1/spec/09First.js]: Description is copy-pasted from the previous test.

**Comment by ilyahumbar** [comment]
- [ ] Reconsider folder structure. Separate your code from 3rd-party dependencies
- [ ] Check with codding standards
 - Braces should be used around all statements, even single statements, when they are part of a control structure, such as an "if"
or "for" statement. This makes it easier to add statements without accidentally introducing bugs.
 - Check empty lines across all files. Generally there should not be more than one empty line for logical separation.
- [ ] Test edge cases (passing null values, undefined, empty lists etc.). e.g. check with memoization task reference.
- [ ] Modularization. Move all functional extensions into separate module. Tasks 7 and 8 should be implemented in separate module which depends on the module with functional extensions. Consider RequireJS

File: week2-javascript/jasmine-standalone-2.4.1/spec/11Memoization.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+        var a = multiplyAndCache(3);
+        var b = multiplyAndCache(3);
+        var c = multiplyAndCache(4);
+
+        expect(a).toEqual(300);
+        expect(b).toEqual(300);
+        expect(c).toEqual(400);
+    });
+});
```
Comment (ilyahumbar) [inline week2-javascript/jasmine-standalone-2.4.1/spec/11Memoization.js]: Reconsider this test: test edge cases and circle references. Test that caching is working as intended (original function should not be called more than once)

File: week2-javascript/jasmine-standalone-2.4.1/spec/08SumOfRandomNumbers.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+    it("It seems to be true", function () {
+
+        var result = getRandomSum();
+
+        expect((10 < result) && (result < 990)).toEqual(true);
+    });
+
+
+});
```
Comment (ilyahumbar) [inline week2-javascript/jasmine-standalone-2.4.1/spec/08SumOfRandomNumbers.js]: this test is useless. Here you should test that your implementation call linearFold function with proper arguments.
Or even I'd suggest to define separate function (sum, which takes a list of numbers) and write tests for it.   
getRandomSum test should verify that function 'getRandomSum' calls function 'sum' in its implementation

File: week2-javascript/jasmine-standalone-2.4.1/src/BogdanWeekTwoAllScripts.js [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+function lazy(f) {
+    var args = Array.prototype.slice.call(arguments, 1);
+    return function () {
+        return f(...args);
+    }
+}
+
+//Problem 11: Memoization
+
+function memoize(f) {
```
Comment (ilyahumbar) [inline week2-javascript/jasmine-standalone-2.4.1/src/BogdanWeekTwoAllScripts.js]: Make this function remember not only the last computation but also all previous computations. Add support of multi-argument function as a parameter

File: week2-javascript/Scripts/FunctionalExtensions.js [hunk 5]
```diff
@@ -110,14 +110,19 @@ function lazy(f) {
...
+        return memoize;
+    }
+    if (f === undefined || f === null) {
+        return undefined;
+    }
+    var memo = {};
+    return function () {
+        var args = Array.prototype.slice.call(arguments);
+        if (args in memo) {
+            return memo[args];
```
Comment (ilyahumbar) [inline week2-javascript/Scripts/FunctionalExtensions.js]: This is not working if arguments contain objects. Write test case for this scenario as well

File: week2-javascript/Scripts/FunctionalExtensions.js [hunk 6]
```diff
@@ -110,14 +110,19 @@ function lazy(f) {
...
         }
-        prevousInput = x;
-        previousResult = f(x);
-        return previousResult;
+        memo[args] = f(...args);
```
Comment (ilyahumbar) [inline week2-javascript/Scripts/FunctionalExtensions.js]: since we don't use transpilers i suggest to avoid ES2015 syntax. Include transiler to ES5 or use ES5 only

File: week2-javascript/Scripts/OtherProblems.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
+'use strict';
+
+console.log('Meow meow');
```
Comment (ilyahumbar) [inline week2-javascript/Scripts/OtherProblems.js]: :)
No debugging code like this in committed files. Also avoid code commenting without special note (it raises question: why the code is commented?)

File: week2-javascript/Tests/spec/08SumOfRandomNumbers.js [hunk 8]
```diff
@@ -1,11 +1,17 @@
...
-        expect((10 < result) && (result < 990)).toEqual(true);
+        expect(result).toEqual(30);
     });
 
+    it("Sum was called", function () {
+        getSum = jasmine.createSpy('getSum');
+
+        var result = getRandomSum();
 
+        expect(getSum).toHaveBeenCalled();
```
Comment (ilyahumbar) [inline week2-javascript/Tests/spec/08SumOfRandomNumbers.js]: Not important, but you also can test here that generateRandomArray() function have been called

**Comment by ilyahumbar** [comment]
- [ ] Use lisp-case (lowercase words separated by dashes) naming for files and folders
- [ ] Use the following folders structure:
  - week2-javascript/vendor (for any 3rd party dependencies)
  - week2-javascript/tests for jasmine tests
  - week2-javascript/ for your source files.
